R/oauth-flow-auth-code.R

```diff
@@ -110,12 +116,16 @@ req_oauth_auth_code <- function(req, client,
 #' @param auth_params List containing additional parameters passed to `oauth_flow_auth_code_url()`
 #' @param token_params List containing additional parameters passed to the
 #'   `token_url`.
-#' @param host_name Host name used to generate `redirect_uri`
-#' @param host_ip IP address web server will be bound to.
-#' @param port Port to bind web server to. By default, this uses a random port.
-#'   You may need to set it to a fixed port if the API requires that the
+#' @param host_name Deprecated. Use `redirect_url` instead.
+#' @param host_ip IP address for the temporary webserver used to capture the
+#'   authorization code.
+#' @param port Port to bind the temporary webserver to. Used only when
```

Comment:
Could we just extract this from the `redirect_url`? I'm actually somewhat surprised that the random port works at all, because I thought the redirect url matching was quite strict.

Overall something no longer feels quite right about the factoring of these arguments, but I'm not quite sure how to change it. 

Maybe we can drop `host_ip` since it seems unlikely that anything other than 127.0.0.1 will work?

Reply:
* Redirect URL matching is _generally_ strict about ports except in the very limited case of `localhost`, which some providers are lax about so that local development is easier (examples that I can find are Keycloak, Pingfederate, GitHub, possibly Azure AD).
* I'm also surprised that `host_ip` exists. I think it may have originally been to allow for `0.0.0.0` so that you could, in theory, do something like access an RStudio Server instance on `rs.acme.internal:1410` with `host_name = "rs.acme.internal", port = 1410, host_ip = "0.0.0.0"`. This would be pretty limited compared to specifying the redirect URL, though, and assumes a lot about firewalls. I've never tried it.

Reply:
Also: one big limitation of the `host_name = "rs.acme.internal", port = 1410, host_ip = "0.0.0.0"` approach is that it can't be used to specify HTTPS, and _most_ OAuth2 providers won't let you redirect to a non-HTTPS URL (other than, again, the special `localhost` case).

Overall it felt like the right approach to just add `redirect_url` as an argument rather than trying to expose enough other knobs that it would get generated correctly.

Reply:
I agree that `redirect_uri` is the right thing to expose; I was wondering if we should move to eliminate some of the old knobs.